drivers/dac/dac_gd32.c

```diff
@@ -0,0 +1,143 @@
+ * But some SOC, like GD32F350 only have one output, DAC0.
+ */
+#if !defined(CONFIG_SOC_GD32F350)
+#define DAC_GD32_OUT_NUM  2
+#define DAC_GD32_OUT_DUAL 1
+#else
+#define DAC_GD32_OUT_NUM  1
+#endif /* !CONFIG_SOC_GD32F350 */
+
+#define DAC_GD32_DEFAULT_VALUE 0
+
+struct dac_gd32_config {
+	uint32_t reg;
+	uint32_t rcu_periph_clock;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+struct dac_gd32_data {
+	uint8_t resolutions[DAC_GD32_OUT_NUM];
+};
+
+static void dac_gd32_write(struct dac_gd32_data *data,
+		    uint8_t dacx, uint32_t value)
+{
+	uint32_t align;
+
+	if (data->resolutions[dacx] == 8) {
+		align = DAC_ALIGN_8B_R;
+	} else {
+		align = DAC_ALIGN_12B_R;
+	}
+
+#ifdef DAC_GD32_OUT_DUAL
+	dac_data_set(dacx, align, value);
+#else
+	dac_data_set(align, value);
+#endif /* DAC_GD32_OUT_DUAL */
+}
+
+static int dac_gd32_channel_setup(const struct device *dev,
+			   const struct dac_channel_cfg *channel_cfg)
+{
+	struct dac_gd32_data *data = dev->data;
+	uint8_t dacx = channel_cfg->channel_id;
+
+	if (dacx >= DAC_GD32_OUT_NUM) {
+		return -ENOTSUP;
+	}
+
+	if (channel_cfg->resolution == 8 ||
+	    channel_cfg->resolution == 12) {
+		data->resolutions[dacx] = channel_cfg->resolution;
+	} else {
+		return -ENOTSUP;
+	}
+
+#if DAC_GD32_OUT_DUAL
+	dac_disable(dacx);
+	dac_gd32_write(data, dacx, DAC_GD32_DEFAULT_VALUE);
+	dac_enable(dacx);
+#else
+	dac_deinit();
+	dac_gd32_write(data, dacx, DAC_GD32_DEFAULT_VALUE);
+	dac_enable();
+#endif /* DAC_GD32_OUT_DUAL */
+
+	return 0;
+}
+
+static int dac_gd32_write_value(const struct device *dev,
+			 uint8_t dacx, uint32_t value)
+{
+	struct dac_gd32_data *data = dev->data;
+
+	if (dacx >= DAC_GD32_OUT_NUM) {
+		return -ENOTSUP;
+	}
+
+	dac_gd32_write(data, dacx, value);
+
+	return 0;
+}
+
+struct dac_driver_api dac_gd32_driver_api = {
+	.channel_setup = dac_gd32_channel_setup,
+	.write_value = dac_gd32_write_value
+};
+
+static int dac_gd32_init(const struct device *dev)
+{
+	const struct dac_gd32_config *cfg = dev->config;
+	int ret;
+
+	ret = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		LOG_ERR("Failed to apply pinctrl state");
+		return ret;
+	}
+
+	rcu_periph_clock_enable(cfg->rcu_periph_clock);
+
+	return 0;
+}
+
+#define DAC_GD32_INIT(idx)							\
+	BUILD_ASSERT((idx) == 0,						\
+		     "multiple instances not supported.");			\
+	PINCTRL_DT_INST_DEFINE(idx)						\
+	static struct dac_gd32_data dac_gd32_data_##idx;			\
+	static const struct dac_gd32_config dac_gd32_cfg_##idx = {		\
+		.reg = DT_INST_REG_ADDR(idx),					\
+		.rcu_periph_clock = DT_INST_PROP(idx, rcu_periph_clock),	\
+		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx)			\
+	};									\
+	DEVICE_DT_INST_DEFINE(0,						\
+			      &dac_gd32_init,					\
+			      NULL,						\
+			      &dac_gd32_data_##idx,				\
+			      &dac_gd32_cfg_##idx,				\
+			      POST_KERNEL,					\
+			      CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,		\
+			      &dac_gd32_driver_api);				\
+
+DT_INST_FOREACH_STATUS_OKAY(DAC_GD32_INIT)
```

Comment:
since DAC is a singleton device, you can just define one instance, no need to use `DT_INST_FOREACH_STATUS_OKAY`

Reply:
done.